### PR TITLE
[Python] Improve performance of conversion to Numpy/Pandas for nested lists

### LINKED ISSUE #8623: High memory usage when reading boolean list column from parquet
### What happens?

Memory usage explodes when I try to materialize a 100k row BOOL[] column(ie each row value is [ False, True....1000 elements]) as a pandas dataframe from parquet files. 
At same time, if I try to materialize the column as a arrow table first and then convert it to a pandas dataframe, memory usage doesn't spike.

### To Reproduce

```
 parquet_path = '<path>'

 query = "select \"boolean_list_column\" from read_parquet('{}/*.parquet')".format(parquet_path)
#memory usage explodes to like 30GB
 column = duckdb.sql(query).df()

#memory usage stays level around 2GB
 column = duckdb.sql(query).arrow().to_pandas()
```

### OS:

mac

### DuckDB Version:

0.8.1

### DuckDB Client:

python

### Full Name:

Paul Cherian

### Affiliation:

qualtrics

### Have you tried this on the latest `master` branch?

I have tested with a master build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by Tishj**
This PR fixes #8623

Previously we were creating python `list` objects, now we create `numpy.array` objects instead.
This has the advantage that we no longer use the slow tuple-at-a-time conversion we used previously, and also in most cases the memory usage for converting nested lists goes down significantly because we make fewer small allocations.

### CODE REVIEW

**Comment by Mytherin** [comment]
Thanks for the PR! Looks great - I played around with this a bit and ran some benchmarks out of curiosity, here are some timings:

|                   Benchmark                    | v0.10.0 | New  |
|------------------------------------------------|---------|------|
| 1M Boolean Lists (1K rows each)                | 82.8s   | 6.7s |
| 1M Integer Lists (1K rows each)                | 101.1s  | 4.9s |
| 100K string lists (1K rows each)               | 21.1s   | 7.6s |
| 100K nested integer lists (10 * 100 rows each) | 16.7s   | 8.1s |


###### Boolean List

```sql
select [x%2==0 for x in generate_series(i, i + 1000, 1)] l from range(1000000) t(i);
```

###### Integer List

```sql
select generate_series(i, i + 1000, 1) l from range(1000000) t(i);
```

###### String List

```sql
select [concat('thisisastring', x) for x in generate_series(i, i + 1000, 1)] l from range(100000) t(i);
```

###### Nested List

```sql
select [
	generate_series(i, i + 100, 1),
	generate_series(i + 100, i + 200, 1),
	generate_series(i + 200, i + 300, 1),
	generate_series(i + 300, i + 400, 1),
	generate_series(i + 400, i + 500, 1),
	generate_series(i + 500, i + 600, 1),
	generate_series(i + 600, i + 700, 1),
	generate_series(i + 700, i + 800, 1),
	generate_series(i + 800, i + 900, 1),
	generate_series(i + 900, i + 1000, 1)
] l from range(100000) t(i);
```

File: tools/pythonpkg/src/numpy/array_wrapper.cpp [hunk 1]
```diff
@@ -274,19 +274,37 @@ struct UUIDConvert {
...
+		D_ASSERT(list_offset + list_size <= child_size);
+		result.Append(0, child_vector, child_size, list_offset, list_size);
+		return result.ToArray();
 	}
 };
 
 struct StructConvert {
-	static py::dict ConvertValue(Vector &input, idx_t chunk_offset, const ClientProperties &client_properties) {
+	static py::dict ConvertValue(Vector &input, idx_t chunk_offset, NumpyAppendData &append_data) {
+		auto &client_properties = append_data.client_properties;
```
Comment (Mytherin) [inline tools/pythonpkg/src/numpy/array_wrapper.cpp:306]: We are losing the source offset in `StructConvert`, `MapConvert`, `UnionConvert`, etc. Will this work correctly when these types are present in lists? Can we add some tests?

Comment (Tishj) [reply tools/pythonpkg/src/numpy/array_wrapper.cpp:306]: I think we are adding it to the `chunk_offset` passed to `ConvertNested` which calls this, but I'll check never the less 👍

File: tools/pythonpkg/tests/fast/pandas/test_pandas_update.py [hunk 2]
```diff
@@ -8,4 +8,6 @@ def test_pandas_update_list(self, duckdb_cursor):
         duckdb_cursor.execute('create table t (l int[])')
         duckdb_cursor.execute('insert into t values ([1, 2]), ([3,4])')
         duckdb_cursor.execute('update t set l = [5, 6]')
-        assert duckdb_cursor.execute('select * from t').fetchdf()['l'].tolist() == [[5, 6], [5, 6]]
+        expected = pd.DataFrame({'l': [[5, 6], [5, 6]]})
+        res = duckdb_cursor.execute('select * from t').fetchdf()
+        pd.testing.assert_frame_equal(expected, res)
```
[the remark this answers is not in the document]
Comment (Tishj) [reply tools/pythonpkg/tests/fast/pandas/test_pandas_update.py:13]: I have no clue what this test is testing, I just fixed it up so it doesn't break with this change
We have `test_fetch_nested` which was being skipped before, I fixed it up now so hopefully it won't cause random CI breakages again (that's apparently why it was skipped before)

[the remark this answers is not in the document]
Comment (Tishj) [reply tools/pythonpkg/tests/fast/pandas/test_pandas_update.py:13]: Seems that the manylinux CI job is still dying on this test

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**